src/components/Synthetics/ChartTokenSelector/ChartTokenSelector.tsx

```diff
@@ -0,0 +1,246 @@
+      const maxLongLiquidity = getAvailableUsdLiquidityForPosition(marketInfo, true);
+      const maxShortLiquidity = getAvailableUsdLiquidityForPosition(marketInfo, false);
+
+      return {
+        maxLongLiquidity,
+        maxShortLiquidity,
+        marketTokenAddress: marketInfo.marketTokenAddress,
+        indexTokenAddress: marketInfo.indexTokenAddress,
+      };
+    });
+    const groupedMarketsWithIndex: { [marketAddress: string]: TokenOption[] } = groupBy(
+      marketsWithMaxReservedUsd,
+      (market) => market.indexTokenAddress
+    );
+
+    return groupedMarketsWithIndex;
+  }, [sortedAllMarkets]);
+
+  function handleMarketSelect(token: Token, maxLongLiquidityPool: TokenOption, maxShortLiquidityPool: TokenOption) {
+    const tokenAddress = convertTokenAddress(chainId, token.address, "wrapped");
+
+    if (tokenAddress === selectedToken?.address) return;
+
+    if (tradeFlags?.isSwap) {
+      onSelect({
+        indexTokenAddress: token.address,
+      });
+      return;
+    }
+
+    const currentExistingPositions = Object.values(positionsInfo || {}).filter((position) => {
+      if (position.isLong === isLong) {
+        return convertTokenAddress(chainId, position.marketInfo.indexTokenAddress, "wrapped") === tokenAddress;
+      }
+      return false;
+    });
+
+    let marketTokenAddress;
+    const largestExistingPosition =
+      Array.isArray(currentExistingPositions) && currentExistingPositions.length
+        ? currentExistingPositions.reduce((max, current) => (max.sizeInUsd.gt(current.sizeInUsd) ? max : current))
+        : undefined;
+
+    if (largestExistingPosition) {
+      marketTokenAddress = largestExistingPosition?.marketInfo.marketTokenAddress;
+    } else {
+      if (isLong) {
+        marketTokenAddress = maxLongLiquidityPool?.marketTokenAddress;
+      }
+
+      if (isShort) {
+        marketTokenAddress = maxShortLiquidityPool?.marketTokenAddress;
+      }
+    }
+
+    onSelect({
+      indexTokenAddress: token.address,
+      marketTokenAddress,
+    });
+  }
+
+  function getMaxLongShortLiquidityPool(token: Token) {
+    const indexTokenAddress = token.isNative ? token.wrappedAddress : token.address;
+    const currentMarkets = groupedIndexMarkets[indexTokenAddress!];
+    const maxLongLiquidityPool = currentMarkets?.reduce((prev, current) => {
+      if (!prev.maxLongLiquidity || !current.maxLongLiquidity) return current;
+      return prev.maxLongLiquidity.gt(current.maxLongLiquidity) ? prev : current;
+    });
+
+    const maxShortLiquidityPool = currentMarkets?.reduce((prev, current) => {
+      if (!prev.maxShortLiquidity || !current.maxShortLiquidity) return current;
+      return prev.maxShortLiquidity.gt(current.maxShortLiquidity) ? prev : current;
+    });
+    return {
+      maxLongLiquidityPool,
+      maxShortLiquidityPool,
+    };
+  }
+
+  return (
+    <Popover className="Synths-ChartTokenSelector">
+      {({ open, close }) => {
+        if (!open) setSearchKeyword("");
```

Comment:
looks dangerous to set state only by "open" field, probably we could also chech that search.length > 0

Reply:
Updated!